Rewrite develop/advanced_plugin_testing (#118)
Limnoria/Limnoria-doc: develop/advanced_plugin_testing.rst

***********************
Advanced Plugin Testing
***********************

Why Write Tests?
================

`Automated testing <https://en.wikipedia.org/wiki/Test_automation>`_ is a
hallmark of modern software development. Adding a test suite to your plugin
makes it easier to update and redesign your code, update its dependencies, etc.
without having to manually check that every feature still works.

Plugin Tests
============

First, we assume that you have already gone through the
:ref:`Plugin Development Tutorial <plugin-tutorial>` and created the plugin
boilerplate with :command:`supybot-plugin-create`.

Plugin Test Case Classes
------------------------

Limnoria comes with two plugin test case classes,
:class:`supybot.test.PluginTestCase` and
:class:`supybot.test.ChannelPluginTestCase`. Use the latter when the plugin's
commands need to be run in a channel, and the former otherwise.

Both of these classes inherit from Python's built-in
`unittest.TestCase <https://docs.python.org/3/library/unittest.html>`_; as such,
most of the documentation there applies to plugin tests too.

Note that instead
of ``python -m unittest``, Limnoria plugin tests are run using the

> :command:`supybot-test` command: e.g. ``supybot-test /path/to/your/Plugin``

Reviewer: ```suggestion
:command:`limnoria-test` command: e.g. ``limnoria-test /path/to/your/Plugin``
```

of your plugin simpler. After all, we want people to be able to run your
plugin tests without having to have all of your plugins!

One last thing to note before moving along is that each of the test methods
should describe what they are testing. If you want to test that your plugin
only responds to registered users, don't be afraid to name your test method
testOnlyRespondingToRegisteredUsers or testNotRespondingToUnregisteredUsers.
You may have noticed some rather long and seemingly unwieldy test method names
in our code, but that's okay because they help us know exactly what's failing
when we run our tests. With an ambiguously named test method we may have to
crack open test.py after running the tests just to see what it is that failed.
For this reason you should also test only one thing per test method. Don't
write a test method named testFoobarAndBaz. Just write two test methods,
testFoobar and testBaz. Also, it is important to note that test methods must
begin with test and that any method within the class that does begin with test
will be run as a test by the supybot-test program. If you want to write
utility functions in your test class that's fine, but don't name them

> something that begins with test or they will be executed as tests.

Reviewer: I think it would be good to keep a paragraph about this

>             # Create a user to send IRC messages from
>             self.prefix = 'foo!bar@baz'

Reviewer: ```suggestion
            # Define the identity of the user who we send messages as
            self.prefix = 'foo!bar@baz'
```

(the comment applied to the whole code block, not just that line)
